File: 이준영/백준/Gold/14725. 개미굴/개미굴.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {

	public static class Node {

		Map<String, Node> child;
		int depth;
		String cur;
		ArrayList<Node> lst;

		public Node() {
			this.child = new HashMap<>();
			this.lst = new ArrayList<>();
			this.cur = "";
			this.depth = 0;
		}

	}

	public static class Trie {
		Node root;

		public Trie() {
			this.root = new Node();
		}

		public void insert(String[] str_arr) {
			int dep = 0;
			int len = str_arr.length;
			Node tmp = this.root;
			for (int i = 0; i < len; i++) {
				dep += 1;
				if (tmp.child.containsKey(str_arr[i])) {
					tmp = tmp.child.get(str_arr[i]);
				} else {
					Node new_node = new Node();
					new_node.cur = str_arr[i];
					tmp.child.put(str_arr[i], new_node);
					tmp.lst.add(new_node);
					tmp = new_node;
				}
				tmp.depth = dep;
			}
		}

		public void print() {

			StringBuilder sb = new StringBuilder();
			Set<Node> visited = new HashSet<>();
			Stack<Node> stack = new Stack<>();
			stack.push(root);

			while (!stack.isEmpty()) {
				Node tmp = stack.pop();
				if (!visited.contains(tmp)) {
					for (int i = 1; i < tmp.depth; i++) {
						sb.append("--");
						
					}
					if(tmp != this.root) {
						sb.append(tmp.cur).append("\n");
					}

					if (tmp.child.isEmpty()) {
						continue;
					} else {
						Collections.sort(tmp.lst, new Comparator<Node>() {
							@Override
							public int compare(Node p1, Node p2) {
								return p2.cur.compareTo(p1.cur);
							}
						});
						for (Node tmp_node : tmp.lst) {
							if (!visited.contains(tmp_node)) {
								stack.push(tmp_node);
							}
						}
					}
				}
			}

			System.out.println(sb);
		}
	}

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		Trie trie = new Trie();
		for (int i = 0; i < N; i++) {
			
			st = new StringTokenizer(br.readLine());
			int K = Integer.parseInt(st.nextToken());
			String[] str_arr = new String[K];
			
			for (int j = 0; j < K; j++) {
				str_arr[j] = st.nextToken();
			}
			
			trie.insert(str_arr);
		}
		
		trie.print();

	}
}

File: 이준영/백준/Gold/14725. 개미굴/개미굴.py
import sys

sys.setrecursionlimit(10 ** 5)
input = sys.stdin.readline


class Node:
    def __init__(self, name=""):
        self.name = name
        self.child = {}  # key: 문자열, value: Node
        self.depth = 0  # 루트 바로 아래 자식이 depth=1


class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, parts):
        node = self.root
        for d in range(len(parts)):
            part = parts[d]
            if part not in node.child:
                new_node = Node(part)
                new_node.depth = d + 1
                node.child[part] = new_node
            node = node.child[part]
            node.depth = d + 1

    def dfs_print(self, node):
        # 자식 키를 사전 순으로 정렬하여 출력
        for key in sorted(node.child):
            child = node.child[key]
            indent = "--" * (child.depth - 1)
            print(f"{indent}{child.name}")
            self.dfs_print(child)

    def print(self):
        self.dfs_print(self.root)


N = int(input().split()[0])
trie = Trie()
for _ in range(N):
    data = input().split()
    k = int(data[0])
    path = data[1:]
    trie.insert(path)
trie.print()
